src/kOS.Safe/Encapsulation/EnumerableValue.cs

```diff
@@ -30,23 +34,44 @@ IEnumerator IEnumerable.GetEnumerator()
 
         public bool Contains(T item)
         {
-            return Collection.Contains(item);
+            return InnerEnumerable.Contains(item);
         }
 
-        public abstract int Count { get; }
+        public int Count()
+        {
+            return InnerEnumerable.Count();
+        }
 
         public override string ToString()
         {
             return new SafeSerializationMgr().ToString(this);
         }
 
+        public override Dump Dump()
+        {
+            var result = new DumpWithHeader
+            {
+                Header = label + " of " + InnerEnumerable.Count() + " items:"
+            };
+
+            int i = 0;
+            foreach (T item in this)
+            {
+                result.Add(i, item);
+                i++;
+            }
+
+            return result;
+        }
+
         private void InitializeEnumerableSuffixes()
         {
-            AddSuffix("ITERATOR",   new NoArgsSuffix<Enumerator>          (() => new Enumerator(Collection.GetEnumerator())));
-            AddSuffix("CONTAINS",   new OneArgsSuffix<BooleanValue, T>    (item => Collection.Contains(item)));
-            AddSuffix("EMPTY",      new NoArgsSuffix<BooleanValue>        (() => !Collection.Any()));
-            AddSuffix("DUMP",       new NoArgsSuffix<StringValue>         (() => new StringValue(ToString())));
+            AddSuffix("ITERATOR",        new NoArgsSuffix<Enumerator>(() => new Enumerator(InnerEnumerable.GetEnumerator())));
+            AddSuffix("REVERSEITERATOR", new NoArgsSuffix<Enumerator>(() => new Enumerator(Enumerable.Reverse(InnerEnumerable).GetEnumerator())));
+            AddSuffix("LENGTH",          new NoArgsSuffix<ScalarValue>(() => InnerEnumerable.Count()));
+            AddSuffix("CONTAINS",        new OneArgsSuffix<BooleanValue, T>((n) => Contains(n)));
+            AddSuffix("EMPTY",           new NoArgsSuffix<BooleanValue>(() => !InnerEnumerable.Any()));
+            AddSuffix("DUMP",            new NoArgsSuffix<StringValue>(() => new StringValue(ToString())));
```

Comment:
Actually this was my change.  I was attempting to compromise.  I do not like the extra white space  that comes with the indentation, but others think it helps to clarify.  So I thought that it would end up being OK if the methods themselves weren't indented, but the constructor was.  I'd rather just eliminate the indentation outright.